File: tools/clod-poc/src/qa/unified/edge_metrics.ts
import { rec709Luminance, type LinearImage } from "./image_linear.js";

export function sobelMagnitudes(image: LinearImage): Float32Array {
  const luminance = new Float32Array(image.width * image.height);
  for (let pixel = 0; pixel < luminance.length; pixel++) {
    const offset = pixel * 3;
    luminance[pixel] = rec709Luminance(image.rgb[offset] ?? 0, image.rgb[offset + 1] ?? 0, image.rgb[offset + 2] ?? 0);
  }
  const output = new Float32Array(luminance.length);
  for (let y = 1; y < image.height - 1; y++) {
    for (let x = 1; x < image.width - 1; x++) {
      const i = y * image.width + x;
      const tl = luminance[i - image.width - 1] ?? 0;
      const tc = luminance[i - image.width] ?? 0;
      const tr = luminance[i - image.width + 1] ?? 0;
      const ml = luminance[i - 1] ?? 0;
      const mr = luminance[i + 1] ?? 0;
      const bl = luminance[i + image.width - 1] ?? 0;
      const bc = luminance[i + image.width] ?? 0;
      const br = luminance[i + image.width + 1] ?? 0;
      const gx = -tl + tr - 2 * ml + 2 * mr - bl + br;
      const gy = -tl - 2 * tc - tr + bl + 2 * bc + br;
      output[i] = Math.hypot(gx, gy);
    }
  }
  return output;
}

File: src/diagnostics/qa/unified/edge_metrics.rs
use super::image_linear::{LinearImage, rec709_luminance};

pub fn sobel_magnitudes(image: &LinearImage) -> Vec<f32> {
    let width = image.width as usize;
    let height = image.height as usize;
    let mut luminance = vec![0.0; width * height];
    for (pixel, output) in luminance.iter_mut().enumerate() {
        let offset = pixel * 3;
        *output = rec709_luminance(
            image.rgb[offset],
            image.rgb[offset + 1],
            image.rgb[offset + 2],
        );
    }

    let mut output = vec![0.0; luminance.len()];
    if width < 3 || height < 3 {
        return output;
    }
    for y in 1..height - 1 {
        for x in 1..width - 1 {
            let index = y * width + x;
            let top_left = luminance[index - width - 1];
            let top_center = luminance[index - width];
            let top_right = luminance[index - width + 1];
            let middle_left = luminance[index - 1];
            let middle_right = luminance[index + 1];
            let bottom_left = luminance[index + width - 1];
            let bottom_center = luminance[index + width];
            let bottom_right = luminance[index + width + 1];
            let gx = -top_left + top_right - 2.0 * middle_left + 2.0 * middle_right
                - bottom_left
                + bottom_right;
            let gy = -top_left - 2.0 * top_center - top_right
                + bottom_left
                + 2.0 * bottom_center
                + bottom_right;
            output[index] = gx.hypot(gy);
        }
    }
    output
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flat_image_has_no_edges() {
        let image = LinearImage {
            width: 3,
            height: 3,
            rgb: vec![0.5; 27],
        };
        assert!(sobel_magnitudes(&image).iter().all(|value| *value == 0.0));
    }
}
